Added per stream stats for CUDA Caching Allocator

c-odrin opened the pull request:
Test Plan: Added test to test_cuda.py

Differential Revision: D41309246

awgu commented:
@zdevito @c-odrin Could we plan to try to land this by early next half? I think the combination of @zdevito's memory snapshots, @zhaojuanmao's upstreamed memory tracker, and these per-stream statistics may go far in debugging memory performance for FSDP but surely also for other use cases.

c-odrin committed:
Added per stream stats for CUDA Caching Allocator (#89806)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/89806

Extended the memory statistics per stream which will allow to troubleshoot memory utilization issues and optimize workloads where multiple streams are being used.

Test Plan: Added test to test_cuda.py

Differential Revision: D41309246

fbshipit-source-id: 22abb3fcf9359fe7b80502cc761ee2388cb9c2f4
